File: content/develop/use-cases/prefetch-cache/redis-py/sync_worker.py
"""
Background sync worker for the prefetch-cache demo.

A daemon thread drains the primary's change queue and applies each event
to Redis through ``PrefetchCache.apply_change``. In a real system, the
queue is replaced by a CDC pipeline (Redis Data Integration, Debezium,
or an equivalent) that tails the primary's binlog/WAL and writes the
same shape of events.

The worker exposes ``pause()`` and ``resume()`` so maintenance paths
(``/reprefetch``, ``clear()``) can stop event application without
tearing the thread down. ``pause()`` blocks until the worker is parked,
so the caller knows no apply is in flight by the time it returns.
"""

from __future__ import annotations

from threading import Event, Thread
from typing import Optional

from cache import PrefetchCache
from primary import MockPrimaryStore


class SyncWorker:
    """Drain primary change events into Redis on a daemon thread."""

    def __init__(
        self,
        primary: MockPrimaryStore,
        cache: PrefetchCache,
        poll_timeout_s: float = 0.05,
    ) -> None:
        self.primary = primary
        self.cache = cache
        self.poll_timeout_s = poll_timeout_s
        self._stop_event = Event()
        self._pause_event = Event()
        self._paused_idle_event = Event()
        self._thread: Optional[Thread] = None

    def start(self) -> None:
        if self._thread is not None and self._thread.is_alive():
            return
        self._stop_event.clear()
        self._pause_event.clear()
        self._paused_idle_event.clear()
        self._thread = Thread(
            target=self._run,
            name="prefetch-cache-sync",
            daemon=True,
        )
        self._thread.start()

    def stop(self, join_timeout_s: float = 2.0) -> None:
        """Signal the worker to exit and join its thread.

        If the join times out the worker is wedged inside ``apply_change``;
        we leave ``self._thread`` populated so a subsequent ``start()`` does
        not spawn a second worker on top of the orphan.
        """
        self._stop_event.set()
        if self._thread is None:
            return
        self._thread.join(timeout=join_timeout_s)
        if not self._thread.is_alive():
            self._thread = None

    def pause(self, timeout_s: float = 2.0) -> bool:
        """Stop applying events and block until the worker is parked.

        Returns ``True`` once the worker has confirmed it is idle, or
        ``False`` if the timeout elapsed first. While paused, change
        events accumulate in the primary's queue and are applied in order
        after ``resume()``.
        """
        self._paused_idle_event.clear()
        self._pause_event.set()
        if self._thread is None or not self._thread.is_alive():
            return True
        return self._paused_idle_event.wait(timeout=timeout_s)

    def resume(self) -> None:
        self._pause_event.clear()
        self._paused_idle_event.clear()

    def _run(self) -> None:
        while not self._stop_event.is_set():
            if self._pause_event.is_set():
                self._paused_idle_event.set()
                # Park until the pause is lifted or the worker is stopped.
                while self._pause_event.is_set() and not self._stop_event.is_set():
                    self._stop_event.wait(timeout=self.poll_timeout_s)
                self._paused_idle_event.clear()
                continue

            change = self.primary.next_change(timeout=self.poll_timeout_s)
            if change is None:
                continue
            try:
                self.cache.apply_change(change)
            except Exception as exc:
                # Demo behaviour: log and drop the event. A production
                # CDC consumer would retry with bounded backoff and
                # expose a dead-letter / error counter; see the guide's
                # "Production usage" section.
                print(f"[sync] failed to apply {change!r}: {exc}")

File: content/develop/use-cases/prefetch-cache/java-lettuce/SyncWorker.java
import java.util.Map;

/**
 * Background sync worker for the prefetch-cache demo.
 *
 * <p>A daemon {@link Thread} drains the primary's change queue and
 * applies each event to Redis through
 * {@link PrefetchCache#applyChange(Map)}. In a real system, the queue
 * is replaced by a CDC pipeline (Redis Data Integration, Debezium, or
 * an equivalent) that tails the primary's binlog/WAL and writes the
 * same shape of events.</p>
 *
 * <p>The worker exposes {@link #pause(long)} and {@link #resume()} so
 * maintenance paths ({@code /reprefetch}, {@link PrefetchCache#clear()})
 * can stop event application without tearing the thread down.
 * {@code pause} blocks until the worker is parked, so the caller knows
 * no apply is in flight by the time it returns.</p>
 */
public class SyncWorker {

    private final MockPrimaryStore primary;
    private final PrefetchCache cache;
    private final long pollTimeoutMs;

    private volatile boolean stopRequested = false;
    private volatile boolean pauseRequested = false;

    // Signals worker has parked itself inside the pause loop. Cleared
    // by the worker on resume so the next pause can wait again.
    private final Object pausedSignal = new Object();
    private volatile boolean pausedIdle = false;

    private Thread thread;

    public SyncWorker(MockPrimaryStore primary, PrefetchCache cache) {
        this(primary, cache, 50L);
    }

    public SyncWorker(MockPrimaryStore primary, PrefetchCache cache, long pollTimeoutMs) {
        if (primary == null || cache == null) {
            throw new IllegalArgumentException("primary and cache are required");
        }
        this.primary = primary;
        this.cache = cache;
        this.pollTimeoutMs = pollTimeoutMs;
    }

    public synchronized void start() {
        if (thread != null && thread.isAlive()) {
            return;
        }
        stopRequested = false;
        pauseRequested = false;
        pausedIdle = false;
        thread = new Thread(this::run, "prefetch-cache-sync");
        thread.setDaemon(true);
        thread.start();
    }

    /**
     * Signal the worker to exit and join its thread.
     *
     * <p>If the join times out the worker is wedged inside
     * {@code applyChange}; we leave {@link #thread} populated so a
     * subsequent {@link #start()} does not spawn a second worker on
     * top of the orphan.</p>
     */
    public synchronized void stop(long joinTimeoutMs) {
        stopRequested = true;
        // Also wake the worker out of any park loop.
        synchronized (pausedSignal) {
            pausedSignal.notifyAll();
        }
        if (thread == null) return;
        try {
            thread.join(joinTimeoutMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return;
        }
        if (!thread.isAlive()) {
            thread = null;
        }
    }

    /**
     * Stop applying events and block until the worker is parked.
     *
     * <p>Returns {@code true} once the worker has confirmed it is idle,
     * or {@code false} if the timeout elapsed first. While paused,
     * change events accumulate in the primary's queue and are applied
     * in order after {@link #resume()}.</p>
     */
    public boolean pause(long timeoutMs) {
        Thread workerSnapshot;
        synchronized (this) {
            workerSnapshot = thread;
        }
        pausedIdle = false;
        pauseRequested = true;
        if (workerSnapshot == null || !workerSnapshot.isAlive()) {
            // No worker running — nothing to wait on, treat as paused.
            return true;
        }
        long deadline = System.nanoTime() + timeoutMs * 1_000_000L;
        synchronized (pausedSignal) {
            while (!pausedIdle && System.nanoTime() < deadline) {
                long remainingNanos = deadline - System.nanoTime();
                if (remainingNanos <= 0) break;
                long remainingMs = Math.max(1L, remainingNanos / 1_000_000L);
                try {
                    pausedSignal.wait(remainingMs);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return pausedIdle;
                }
            }
        }
        return pausedIdle;
    }

    public void resume() {
        pauseRequested = false;
        pausedIdle = false;
        synchronized (pausedSignal) {
            pausedSignal.notifyAll();
        }
    }

    private void run() {
        while (!stopRequested) {
            if (pauseRequested) {
                synchronized (pausedSignal) {
                    // Park until pause is lifted or worker is stopped.
                    // Re-announce "pausedIdle" on every iteration so a
                    // *new* pause() that arrives while we are still
                    // parked from the previous cycle gets acknowledged
                    // immediately, not after the caller's full
                    // pause-timeout.
                    while (pauseRequested && !stopRequested) {
                        pausedIdle = true;
                        pausedSignal.notifyAll();
                        try {
                            pausedSignal.wait(pollTimeoutMs);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            return;
                        }
                    }
                    pausedIdle = false;
                }
                continue;
            }

            Map<String, Object> change = primary.nextChange(pollTimeoutMs);
            if (change == null) {
                continue;
            }
            try {
                cache.applyChange(change);
            } catch (Exception exc) {
                // Demo behaviour: log and drop the event. A production
                // CDC consumer would retry with bounded backoff and
                // expose a dead-letter / error counter; see the guide's
                // "Production usage" section.
                System.err.printf("[sync] failed to apply %s: %s%n",
                        change, exc.getMessage());
            }
        }
    }
}
